### PR TITLE
Refactor TerminalApp and Add Tests for Xaml Content

### LINKED ISSUE #1042: Refactor pure C++ types from TerminalApp.dll into a lib
Right now we have one big dll that is declaring both cppwinrt types (e.g. `AppKeyBindings`, `App`) and pure c++ types (`Tab`, `Profile`, etc.).

If one were to try and create a unittesting project for the TerminalApp project, you unfortunately would be unable to do this, because many of the types are only linked into this TerminalApp.dll. For pure cppwinrt types, this isn't that big of a deal, because instantiating the types just does cppwinrt magic to activate them. However, for pure cpp classes, we can't actually link them. This means we can't test things like json parsing, Pane recursing, settings synthesis. 

So we'll have to pull all the types that aren't winrt types out into their own lib, and have the TerminalApp project consume that lib. 

I'm not worried about TermControl and TermConnection so much, since those are only using cppwinrt types at the moment. Should they get more elaborate, then it might be necessary to do a similar refactoring.

Worries:
1. Does this mean we're going to have now **two** `pch.pch` files for the TerminalApp, one for the lib and one for the dll? could we avoid that? See #946.
2. The lib seems to need to include `winrt/coroutine.h` (and any other cppwinrt headers that components might need). Those are generated files. Right now they're generated by the TerminalApp(dll) build, but theoretically they'll need to be created before that, because the lib will theoretically be built before the dll. This may mean that the lib needs to have some sort of cppwinrt tooling turned on, but I don't know what quite yet.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

* Refactors TerminalApp into two projects: 
  - TerminalAppLib, which builds a .lib, and includes all the code
  - TerminalApp, which builds a dll by linking the lib
* Adds a TerminalApp.Unit.Tests project
  - Includes the ability to test cppwinrt types we've authored using a SxS manifest for unpackaged winrt activation
  - includes the ability to test types with XAML content using an appxmanifest
* Adds a giant doc explaining how this was all done. Really, just go read that doc, it'll really help you understand what's going on in this PR.

## PR Checklist
* [x] Closes #1042 
* [x] Tests added/passed 
  - omg I can check this box for real
* [x] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments
Go read  [doc/cascadia/Unittesting-CppWinRT-Xaml.md](https://github.com/microsoft/terminal/compare/master...dev/migrie/b/add-test-for-xaml-content#diff-4ffb266eae4c692a91dad17e707b436c)

## Validation Steps Performed
Manually ran the tests. I'm really hoping they run in CI too 🤞

### CODE REVIEW

File: OpenConsole.sln [hunk 1]
```diff
@@ -71,7 +71,7 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "Host.Tests.Unit", "src\host
 		{06EC74CB-9A12-429C-B551-8562EC954747} = {06EC74CB-9A12-429C-B551-8562EC954747}
 	EndProjectSection
 EndProject
-Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "TextBuffer.Unit.Tests", "src\buffer\out\ut_textbuffer\TextBuffer.Unit.Tests.vcxproj", "{531C23E7-4B76-4C08-8BBD-04164CB628C9}"
```
Comment (zadjii-msft) [inline OpenConsole.sln]: No idea where this came from. I'm going to let the CI spin a build before I push a fix for this.

File: doc/cascadia/Unittesting-CppWinRT-Xaml.md [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+going to cover the steps that I took to get the Windows Terminal updated to be
+able to test not only our C++/WinRT components, but also pure c++ classes that
+were used in the application, and components that used XAML UI elements.
+
+## Move the C++/WinRT implementation to a static lib
+
+By default, most (newly authored) C++/WinRT components are authored as a dll
+that can be used to activate your types. However, you might have other classes
+in that binary that you want to be able to test, which aren't winrt types. If
+the implenmentation is stitting in a DLL, it'll be hard to write a TAEF unittest
```
Comment (carlos-zamora) [inline doc/cascadia/Unittesting-CppWinRT-Xaml.md]: >stitting [](start = 23, length = 8)

typo

**Comment by zadjii-msft** [comment]
Status update: This is testing a scenario that's actually _NOT SUPPORTED_ by the TAEF VsTest adapter currently. We're working with the TAEF team on getting that fixed. 

Depending on their timeline, we might want to comment out the XAML test for now, to get the rest of the refactoring in.

File: src/cascadia/ut_app/JsonTests.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        Json::Value root;
+        std::string errs;
+        const bool parseResult = reader->parse(content.c_str(), content.c_str() + content.size(), &root, &errs);
+        VERIFY_IS_FALSE(parseResult);
+    }
+
+    void JsonTests::ParseInvalidJson()
+    {
+        const std::string badJson000{ "{ foo : bar : baz }" };
```
Comment (adiviness) [inline src/cascadia/ut_app/JsonTests.cpp]: What's the reason for `000`?

File: tools/runut.cmd [hunk 4]
```diff
@@ -11,4 +11,5 @@ call %TAEF% ^
     %OPENCON%\bin\%PLATFORM%\%_LAST_BUILD_CONF%\ConParser.Unit.Tests.dll ^
     %OPENCON%\bin\%PLATFORM%\%_LAST_BUILD_CONF%\ConAdapter.Unit.Tests.dll ^
     %OPENCON%\bin\%PLATFORM%\%_LAST_BUILD_CONF%\Types.Unit.Tests.dll ^
+    %OPENCON%\bin\%PLATFORM%\%_LAST_BUILD_CONF%\UnitTests_TerminalApp\Terminal.App.Unit.Tests.dll ^
```
Comment (adiviness) [inline tools/runut.cmd]: new tests should also be added to /tools/tests.xml and have the OpenConsole.psm1 testing function updated if necessary.

File: src/cascadia/ut_app/JsonTests.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
+/*
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/ut_app/JsonTests.cpp]: Yea I don't know where that came from. Copied from somewhere, but idk where.

**Comment by simonbuchan** [comment]
Just a note, the doc link is broken, [here's a permalink](https://github.com/microsoft/terminal/blob/babfe0cb88222798ede7e5eeca6c61e78f17696c/doc/cascadia/Unittesting-CppWinRT-Xaml.md)

**Comment by zadjii-msft** [comment]
@miniksa @DHowett-MSFT @carlos-zamora This works again, so hopefully we can get it in after the long weekend

File: doc/cascadia/Unittesting-CppWinRT-Xaml.md [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+kind enough to prototype a fix for us, which is the version
+`10.38.190610001-uapadmin`, which we're using in this repo until an official
+release with the fix is available.
+
+## Move the C++/WinRT implementation to a static lib
+
+By default, most (newly authored) C++/WinRT components are authored as a dll
+that can be used to activate your types. However, you might have other classes
+in that binary that you want to be able to test, which aren't winrt types. If
+the implenmentation is sitting in a DLL, it'll be hard to write a TAEF unittest
```
Comment (carlos-zamora) [inline doc/cascadia/Unittesting-CppWinRT-Xaml.md]:
```suggestion
the implementation is sitting in a DLL, it'll be hard to write a TAEF unittest
```

File: doc/cascadia/Unittesting-CppWinRT-Xaml.md [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+### Create the static lib project
+
+We'll start by creating a new static lib project. The easiest way to do this is
+by copying your existing dll `vcxproj` file into a new file. Make sure to change
+the `ProjectGuid` and to add the new project to your `.sln` file. Then, change
+the `ConfigurationType` to `StaticLibrary`. This Lib should be responsible for
+building all of your headers, `.cpp` files, `.idl`s for your winrt types, and
+any `.xaml` files you might have.
+
+You'll likely need to place this new file into a seperate directory from the
```
Comment (carlos-zamora) [inline doc/cascadia/Unittesting-CppWinRT-Xaml.md]:
```suggestion
You'll likely need to place this new file into a separate directory from the
```

File: doc/cascadia/Unittesting-CppWinRT-Xaml.md [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+      <AdditionalLibraryDirectories>$(SolutionDir)\$(Platform)\$(Configuration)\TerminalAppLib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
+
+      <AdditionalDependencies>TerminalAppLib.lib;%(AdditionalDependencies)</AdditionalDependencies>
+    </Link>
+  </ItemDefinitionGroup>
+```
+
+We are NOT adding a reference to the static lib project's .winmd here. As of the
+2.0.190605.7 CppWinRT nuget package, this is enough for MsBuild and Visual
+Studio to be ale to determine that the static lib's  `.winmd` should be included
```
Comment (carlos-zamora) [inline doc/cascadia/Unittesting-CppWinRT-Xaml.md]:
```suggestion
Studio to be able to determine that the static lib's  `.winmd` should be included
```

File: doc/cascadia/Unittesting-CppWinRT-Xaml.md [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+in this package.
+
+At this point, you might have some mdmerge errors, which complain about
+duplicate types in one of your dependencies. This might especially happen if one
+of your dependencies (ex `A.dll`) is also a dependency for one of your _other_
+dependencies (ex `B.dll`). In this example, your final output project `C.dll`
+depends on both `A.dll` and `B.dll`, and `B.dll` _also_ depends on `A.dll`. If
+you're seeing this, I recommend adding `Private=false` and
+`CopyLocalSatelliteAssemblies=false` to your dependant dlls. In this example,
```
Comment (carlos-zamora) [inline doc/cascadia/Unittesting-CppWinRT-Xaml.md]:
```suggestion
`CopyLocalSatelliteAssemblies=false` to your dependent dlls. In this example,
```

File: doc/cascadia/Unittesting-CppWinRT-Xaml.md [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+To be able to instantiate your C++/WinRT types in a TAEF unittest, you'll need
+to rely on a new feature to Windows in version 1903 which enables unpackaged
+activation of WinRT types. To do this, we'll need to author a SxS manifest that
+lists each of our types, and include it in the dll, and also activate it
+manually from TAEF.
+
+#### Creating the manifest
+
+First, you need to create a manifest file that lists each dll your test depends
+upon, and each of the types in that dll. For example, here's an exerpt from the
```
Comment (carlos-zamora) [inline doc/cascadia/Unittesting-CppWinRT-Xaml.md]:
```suggestion
upon, and each of the types in that dll. For example, here's an excerpt from the
```

File: src/cascadia/TerminalApp/pch.h [hunk 11]
```diff
@@ -7,49 +7,6 @@
...
-TRACELOGGING_DECLARE_PROVIDER(g_hTerminalWin32Provider);
-#include <telemetry\ProjectTelemetry.h>
-#include <TraceLoggingActivity.h>
-
-// JsonCpp
-#include <json.h>
-
-#include <shellapi.h>
-#include <filesystem>
+// This file can be empty - the pch.h in TerminalApp/lib does the heavy lifting
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/pch.h]: If this file is just gonna be empty, could/should we remove it?

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/pch.h]: Unfortunately, MsBuild needs the pch to exist for any cppwinrt projects. It's going to autogenerate other files that depend on including it, even if it's empty #bydesign

**Comment by zadjii-msft** [comment]
> 1. What tests can/should we be adding after this PR goes in? I noticed some are commented out because CI doesn't support them yet.

I'll make another PR to uncomment those tests once this on is in, but hopefully we'll be able to write tests for the TerminalApp stuff that includes and winrt/xaml components. Case in point, I want to add some tests for tracking focus when you close panes, or when you open them, or try to move focus across panes. So, they're not really UI tests, but tests for components that include UI elements.

> 2. Since we'll now have TerminalAppLib and TerminalApp, for the most part, any changes should go in TerminalAppLib? What kind of changes should go into TerminalApp then?

Everything should go into TerminalAppLib now. TerminalApp only exists as a dll wrapper around TerminalAppLib. IIRC I added comments throughout the TerminalApp.vcxproj file saying "DON'T ADD NEW FILES HERE - PUT THEM IN TerminalAppLib INSTEAD". Hopefully, it won't need _any_ changes.

File: src/cascadia/TerminalApp/AppKeyBindingsSerialization.h [hunk 12]
```diff
@@ -2,7 +2,7 @@
 // Licensed under the MIT license.
 //
 // Module Name:
-// - Profile.hpp
+// - AppKeyBindingsSerialization.hpp
```
Comment (miniksa) [inline src/cascadia/TerminalApp/AppKeyBindingsSerialization.h]:
```suggestion
// - AppKeyBindingsSerialization.h
```

File: src/cascadia/TerminalApp/TerminalApp.vcxproj [hunk 13]
```diff
@@ -18,85 +19,28 @@
...
   <ItemGroup>
-    <!-- HERE BE DRAGONS:
-      For any types that use XAML information, if their .idl, .cpp, and .h
-      aren't all marked DependentUpon the corresponding .xaml file,
-      XamlTypeInfo.g.h won't pick it up correctly.
-    -->
-    <ApplicationDefinition Include="App.xaml">
-      <SubType>Designer</SubType>
-    </ApplicationDefinition>
+    <!-- DON'T PUT XAML FILES HERE! Put them in TerminalAppLib.vcxproj -->
```
Comment (miniksa) [inline src/cascadia/TerminalApp/TerminalApp.vcxproj]: Just FYI, people doing things in Visual Studio with the click/drag aren't going to see this... but hopefully we'll catch it on PR.

File: src/cascadia/ut_app/SettingsTests.cpp [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+using namespace Microsoft::Console;
+using namespace TerminalApp;
+using namespace WEX::Logging;
+using namespace WEX::TestExecution;
+
+namespace TerminalAppUnitTests
+{
+    // Unfortunately, these tests _WILL NOT_ work in our CI, until we have a lab
+    // machine available that can run Windows version 18362. Until then, these
+    // tests will be commented out. GH#1012 should move our CI to that version.
```
Comment (miniksa) [inline src/cascadia/ut_app/SettingsTests.cpp]: I finished 1012, but I don't think server 2019 is 18362 yet... you should double check though.

Comment (miniksa) [reply src/cascadia/ut_app/SettingsTests.cpp]: Also, instead of leaving the tests commented out, you could wrap them with a function/helper/block that checks the OS version and returns the TAEF test-skipped status (much like the ones that I wrote that do that for OneCore editions). Then at least they'll run on some developer desktops and light up automatically when the test OS supports it.

File: tools/tests.xml [hunk 15]
```diff
@@ -3,6 +3,7 @@
   <test name="host" type="unit" binary="Conhost.Unit.Tests.dll" />
   <test name="textBuffer" type="unit" binary="TextBuffer.Unit.Tests.dll" />
   <test name="terminalCore" type="unit" binary="Terminal.Core.Unit.Tests.dll" />
+  <test name="terminalApp" type="unit" binary="UnitTests_TerminalApp\Terminal.App.Unit.Tests.dll" />
```
Comment (miniksa) [inline tools/tests.xml]: Why is this one not binplaced where the others are? Did you confirm the CI can see it in the subdirectory?

Comment (zadjii-msft) [reply tools/tests.xml]: Because cppwinrt reasons. All cppwinrt projects get binplaced into their own directory, so that they don't get re-included into other projects. I tried last year to get them to all binplace into `bin/` and it was horrifying. The tests definitely _do_ run in CI though, so that's good!

File: src/cascadia/ut_app/SettingsTests.cpp (hunk 14 above)
Comment (zadjii-msft) [reply src/cascadia/ut_app/SettingsTests.cpp]: So after I dug into this, it turns out we can't preempt this test's deployment. The CI will try deploying the appx it runs in _before_ executing any test code. The CI is running `10.0.17763.557` currently. When it trys deploying the package, it'll fail, before we have the chance to say "please skip me", because the package itself is marked for 18295.
#wontfix

File: src/cascadia/TerminalApp/TerminalApp.vcxproj (hunk 13 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalApp.vcxproj]: Yea I got chu. Hopefully this is enough of a warning for us to be able to see it. Even if they do add it to that project in VS, it's likely to fail spectacularly, so hopefully people will catch it before pushing.

File: src/cascadia/TerminalApp/TerminalApp.vcxproj [hunk 16]
```diff
@@ -151,19 +96,14 @@
       <AdditionalIncludeDirectories>$(OpenConsoleDir)\dep\jsoncpp\json;%(AdditionalIncludeDirectories);</AdditionalIncludeDirectories>
     </ClCompile>
     <Link>
+      <!-- Manually link with the TerminalAppLib.lib we've built. -->
+      <AdditionalLibraryDirectories>$(_BinRoot)\TerminalAppLib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/TerminalApp.vcxproj]: i thought `ProjectReference`ing a static lib did this automatically.

File: src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+      Setting this will tell cppwinrt.build.post.props to copy our output from
+      the default OutDir up one level, so the wapproj will be able to find it.
+     -->
+    <NoOutputRedirection>true</NoOutputRedirection>
+  </PropertyGroup>
+
+  <Import Project="$(OpenConsoleDir)src\common.build.post.props" />
+  <Import Project="$(OpenConsoleDir)src\cppwinrt.build.post.props" />
+
+  <!-- Manually include MUX here, instead of importing its targets file. We need
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj]: oh. oh no. i hate this so much.

File: src/cascadia/TerminalApp/packages.config [hunk 18]
```diff
@@ -2,5 +2,5 @@
 <packages>
   <package id="Microsoft.Toolkit.Win32.UI.XamlApplication" version="6.0.0-preview6.2" targetFramework="native" />
   <package id="Microsoft.UI.Xaml" version="2.2.190611001-prerelease" targetFramework="native" />
-  <package id="Microsoft.Windows.CppWinRT" version="2.0.190417.3" targetFramework="native" />
+  <package id="Microsoft.Windows.CppWinRT" version="2.0.190605.7" targetFramework="native" />
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/packages.config]: make sure you rev this for _all projects_

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 19]
```diff
@@ -665,9 +665,11 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
             }
 
-            const auto modifiers = args.KeyModifiers();
-            const auto altEnabled = WI_IsFlagSet(modifiers, VirtualKeyModifiers::Menu);
-            const auto shiftEnabled = WI_IsFlagSet(modifiers, VirtualKeyModifiers::Shift);
+            const auto modifiers = static_cast<uint32_t>(args.KeyModifiers());
+            // static_cast to a uint32_t because we can't use the WI_IsFlagSet
+            // macro directly with a VirtualKeyModifiers
+            const auto altEnabled = WI_IsFlagSet(modifiers, static_cast<uint32_t>(VirtualKeyModifiers::Menu));
+            const auto shiftEnabled = WI_IsFlagSet(modifiers, static_cast<uint32_t>(VirtualKeyModifiers::Shift));
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalControl/TermControl.cpp]: worried about unrelated code cahnges

File: src/cascadia/WindowsTerminal/WindowsTerminal.vcxproj [hunk 20]
```diff
@@ -1,145 +1,168 @@
...
-    <PropertyPageSchema Include="$(VCTargetsPath)$(LangID)\debugger_local_windows.xml" />
-  </ItemGroup>
-  <PropertyGroup Label="Globals">
-    <DebuggerFlavor>WindowsLocalDebugger</DebuggerFlavor>
-  </PropertyGroup>
-  <PropertyGroup>
-    <!-- DON'T REDIRECT OUR OUTPUT -->
-    <NoOutputRedirection>true</NoOutputRedirection>
-  </PropertyGroup>
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (DHowett-MSFT) [inline src/cascadia/WindowsTerminal/WindowsTerminal.vcxproj]: this whole file changed?

File: src/cascadia/ut_app/SettingsTests.cpp [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+    // When that happens, these tests can be re-added.
+
+    // class SettingsTests
```
Comment (DHowett-MSFT) [inline src/cascadia/ut_app/SettingsTests.cpp]: i honestly prefer big `#if 0` blocks over a bunch of `//` lines

File: src/cascadia/ut_app/TerminalApp.UnitTests.vcxproj [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+      <!-- Manually copy the Windows Terminal manifest to our project directory,
+      so we only need to have one version for the entire solution. -->
+      <Command>
+        (xcopy /Y &quot;$(OpenConsoleDir)src\cascadia\WindowsTerminal\WindowsTerminal.manifest&quot; &quot;$(OpenConsoleDir)src\cascadia\ut_app\TerminalApp.Unit.Tests.manifest*&quot; )
+      </Command>
+    </PreBuildEvent>
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup>
+    <PostBuildEvent>
+      <!-- Manually copy A few needed things to our outdir.
```
Comment (DHowett-MSFT) [inline src/cascadia/ut_app/TerminalApp.UnitTests.vcxproj]: is there really no better way to do this than xcopy? like, `<Content>`?

File: src/cascadia/WindowsTerminal/WindowsTerminal.vcxproj (hunk 20 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/WindowsTerminal.vcxproj]: It _really_ didn't, but for whatever reason git thinks it did. I'm not sure what went wrong.

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 19 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/TermControl.cpp]: Ah this was because of a cppwinrt update. We can't just `WI_IsFlagSet` the `VirtualKeyModifiers` enum anymore.

Comment (zadjii-msft) [reply src/cascadia/TerminalControl/TermControl.cpp]: #bydesign

File: src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj (hunk 17 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj]: literally everything is horrible

File: src/cascadia/TerminalApp/TerminalApp.vcxproj (hunk 16 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalApp.vcxproj]: Hmm. That one might have been left behind. There was a time in the life of this PR it was important, but no longer it seems.
